picard/util/pipe.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from sys import platform as PLATFORM
+
+
+if PLATFORM == "win32" or PLATFORM == "cygwin":
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+    PLATFORM = "windows"
+
+
+class Pipe:
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    NOT_FOUND_MESSAGE: str = "FIFO doesn't exist"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+
+    def __init__(self, app_name: str, app_version: str, args=[]):
+        self.__app_name: str = app_name
+        self.__app_version: str = app_version
+        self.__platform: str = PLATFORM
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+
+        # test if pipe is listened to even if no args provided
+        if type(args) == list:
+            if len(args) == 0:
+                args.append(self.MESSAGE_TO_IGNORE)
+        else:
+            raise ValueError("args argument MUST be a list")
+
+        if self.__platform == "windows":
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            if self.__pipe_exists():
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        os.unlink(self.path)
+                        self.__create_unix_pipe()
+                        break
+            else:
+                self.__create_unix_pipe()
+
+    def __pipe_exists(self) -> bool:
+        return os.path.exists(self.path)
+
+    def __generate_filename(self) -> str:
+        prefix: str = ""
+        username: str = os.getlogin()
+        if self.__platform == "windows":
+            prefix = "\\\\.\\pipe\\"
+        else:
+            prefix = "/tmp/"
```

Comment:
In Windows pipes aren't just files on disk like on Unix, so I'll keep discussing the latter one.

I sometimes see apps using /tmp and IMO it's okay and keeps the logic as simple as possible (the OS deletes the file for us in some cases). But if you insist on changing it I think we could also try the [tempfile library](https://docs.python.org/3/library/tempfile.html)

Reply:
I agree to not use ` /tmp`, especially after what we had discovered recently with macOS removing untouched files from `/tmp`  every few days (PICARD-2459).

Not sure about the best location, though. `~/.config`  seems wrong, and we already have these tasks to improve using the proper directory for some stored data.

IMHO on Linux something under `~/.local`  would make sense. On my system pipewire uses `~/.local/state/pipewire` for some stuff, so maybe `~/.local/state/picard`  would be a possibility.

But it does not fit perfectly, according to [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

>  The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:
>
> - actions history (logs, history, recently used files, …)
> - current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

A system wide service might place such pipes in `/var`, and Flatpak also uses `~/.var`, so this already exists on many systems. But so far this really is only used for Flatpak.

https://stackoverflow.com/questions/15794696/what-is-the-proper-place-to-put-named-pipes-on-linux suggests `/run/user/<userid>/`, and indeed this is used by several applications on my system. But:

- There are Linux setups without system.d, and I'm not really keen on introducing the complexity of detecting this
- There are non-Linux Unixes, e.g. the BSDs (or even Haiku, but that could be patched separately)

In any case we need to handle macOS as a specific case. There something in `~/Library/Application Support/` seems to be the correct place, see https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/MacOSXDirectories/MacOSXDirectories.html

Reply:
> https://stackoverflow.com/questions/15794696/what-is-the-proper-place-to-put-named-pipes-on-linux suggests `/run/user/<userid>/`, and indeed this is used by several applications on my system. But:
> 
>     * There are Linux setups without system.d, and I'm not really keen on introducing the complexity of detecting this

I'm using openRC and I have `$XDG_RUNTIME_DIR` (`/run/user/<userid>/`) in my filesystem, so maybe it's not that init-dependent?

Reply:
That would be good. No idea about the BSDs, though. But maybe if we could unify it on Linux it would be great already.

We could also have a simple fallback if /run/user/<userid> is not available or not writable fall back to ~/.config

Reply:
> That would be good. No idea about the BSDs, though. But maybe if we could unify it on Linux it would be great already.
> 
> We could also have a simple fallback if /run/user/ is not available or not writable fall back to ~/.config

We should also check at least runit and SysV tho, [the list of alternative inits](https://nosystemd.org/) is pretty long but they seem to be the most frequently used ones.

Reply:
Okay, I've been thinking about it and `os.getenv("XDG_RUNTIME_DIR")` with appropriate fallback (in case of lack of such file or permission errors mentioned by you) to ~/.config/MusicBrainz/Picard/pipes (or sth similar in the picard ~/.config folder) seems like the best approach to me.

Reply:
> In any case we need to handle macOS as a specific case. There something in `~/Library/Application Support/` seems to be the correct place, see https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/MacOSXDirectories/MacOSXDirectories.html

@phw is there already a Picard-specific folder on Mac in this destination? If yes, how's it called? If no, how should I call it?

Basically the code to add is (`self.__is_mac` is also imported from constants): 

```python
elif self.__is_mac:
    prefix = os.path.expanduser(f"~/Library/Application Support/{picard_folder_name}/") 
```

Reply:
Also, ~/.config doesn't seem to be the best fallback option, see: 

https://tickets.metabrainz.org/browse/PICARD-2252

Reply:
Picard does not yet create a folder in `~/Library/Application Support/` . It does something in `~/Library/Application Support/Preferences` I think, but that's technically not even correct because macOS expects plist files there. But it's a mess really, we need to redo this and place things (config, cache, plugins) in proper location per platform. But it's a bit ugly thing to handle, because of the testing needed on different plaforms and code needed to handle old locations and data migration. So I'd like to see this happen for Picard 3.

Anyway, for the pipe we should start doing something correct. According to the Apple docs the folder name inside "Application Support" should be the app identifier, that would be "org.musicbrainz.Picard" , there is a constant `PICARD_APP_ID` for this.

Regarding `~/.config`: The user not having permission to access their own config folder should be considered a misconfiguration. It should still be handled better for the config file. But for the pipe I think we can use it as a fallback and we need to handle the case of not being able to create the pipe anyway (falling back to new instance mode).

Reply:
Sorry, I missed this comment yesterday, now we're using `PICARD_APP_ID` on Mac :)